SYNTHESIZER: Use only symbols from the original goto as terminals

qinheping opened the pull request:
Collecting variables from trace will also collect those variables added during loop transformation, such as the car variables. Those variables should not appear in the loop contracts. This commit erases them from the terminals of the enumerating grammar used by the synthesizer.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -244,6 +250,21 @@ enumerative_loop_contracts_synthesizert::compute_dependent_symbols(
+  {
+    if(original_symbol_table.lookup(it->get_identifier()) == nullptr)
+    {
+      auto tmp = it;
+      tmp++;
+      result.erase(it);
+      it = tmp;
```
Wouldn't `it = result.erase(it);` be sufficient here?

tautschnig reviewed (approved):
Is it possible to include a test that demonstrates what difference this change makes?

qinheping commented:
> Is it possible to include a test that demonstrates what difference this change makes?

The only difference is the performance of the enumeration. For example, it took 38 seconds to run the benchmark `synthesizer_04`
https://github.com/diffblue/cbmc/actions/runs/6506277751/job/17671527177?pr=7946#step:13:10913. 
With this PR, the running time goes down to 2 seconds: https://github.com/diffblue/cbmc/actions/runs/6580619258/job/17878990179?pr=7970#step:13:10787.

qinheping committed:
Use only symbols from the original goto as terminals in synthesizer

Collecting variables from trace will also collect those variables added during loop transformation, such as the car variables.
Such variables should not appear in the loop contracts.
This commit erases them from the terminals of the enumearting grammar used by the synthesizer.
